theano/tensor/nnet/ctc_wrapper.c

```diff
@@ -0,0 +1,292 @@
+        free( context->workspace );
+
+    if ( NULL != context->input_lengths )
+        free( context->input_lengths );
+
+    if ( NULL != context->flat_labels )
+        free( context->flat_labels );
+
+    if ( NULL != context->label_lengths )
+        free( context->label_lengths );
+
+    Py_XDECREF( context->activations_copy );
+
+    // Decrease reference to allocated costs and gradients, if an error occurs
+    if ( error )
+    {
+        Py_XDECREF( context->costs );
+        Py_XDECREF( context->gradients );
+    }
+}
+
+int ctc_check_result(ctcStatus_t retcode, const char * msg)
+{
+    if( CTC_STATUS_SUCCESS != retcode )
+    {
+        // Get error message from underlying library
+        const char * ctc_msg = ctcGetStatusString( retcode );
+
+        PyErr_Format( PyExc_RuntimeError,
+                      "%s | CTC library error message: %s",
+                      msg,
+                      ctc_msg );
+        return 1;
+    }
+    return 0;
+}
+
+void create_contiguous_input_lengths( PyArrayObject * input_lengths_arr,
+    int ** input_lengths )
+{
+    int num_elements = PyArray_DIMS( input_lengths_arr )[0];
+
+    *input_lengths = (int *) malloc( num_elements * sizeof(int) );
+
+    if ( NULL == (*input_lengths) )
+        return;
+
+    for( int elem_idx = 0; elem_idx < num_elements; ++elem_idx )
+    {
+        (*input_lengths)[elem_idx] = *( (int *) PyArray_GETPTR1( input_lengths_arr, elem_idx ) );
+    }
+}
+
+void create_flat_labels( PyArrayObject * label_matrix, int ** flat_labels,
+    int ** label_lengths )
+{
+    int rows = PyArray_DIMS( label_matrix )[0];
+    int cols = PyArray_DIMS( label_matrix )[1];
+
+    *flat_labels = (int *) malloc( rows * cols * sizeof(int) );
+    if ( NULL == (*flat_labels) )
+        return;
+
+    *label_lengths = (int *) malloc( rows * sizeof(int) );
+    if ( NULL == (*label_lengths) )
+    {
+        free( *flat_labels );
+        *flat_labels = NULL;
+        return;
+    }
+
+    int label_index = 0;
+    for( int row_idx = 0; row_idx < rows; ++row_idx )
+    {
+        int label_length = 0;
+        for( int col_idx = 0; col_idx < cols; ++col_idx )
+        {
+            int label = *( (int *) PyArray_GETPTR2( label_matrix, row_idx, col_idx ) );
+            if ( label >= 0 )  // negative values are assumed to be padding
+            {
+                (*flat_labels)[ label_index++ ] = label;
+                ++label_length;
+            }
+        }
+        (*label_lengths)[ row_idx ] = label_length;
+    }
+}
+
+#section support_code_apply
+
+int APPLY_SPECIFIC(ctc_cost_cpu)(PyArrayObject *  in_activations,
+                                 PyArrayObject *  in_labels,
+                                 PyArrayObject *  in_input_lengths,
+                                 PyArrayObject ** out_costs,
+                                 PyArrayObject ** out_gradients)
+{
+    ctc_context_t * context = (ctc_context_t *)malloc( sizeof( ctc_context_t ) );
```

Comment:
It may be more convenient to declare context on stack instead of allocating it (on heap). Allocation (malloc) implies deallocation (free), so that you will be forced to call `free(context)` after `ctc_context_destroy(context)` everywhere you want to return.

Instead, you can replace this line with:
```C
// Object on stack.
ctc_context_t context_object;
/*
Pointer to object on stack
(with this, you won't need to replace `context` by `&context` in rest of code).
*/
ctc_context_t* context = &context_object;
```

Then, you won't need to call `free(context)` anywhere, as `context_object` itself will be automatically destroyed by C (because it is on stack).

You still have to call `ctc_context_destroy()` to free internal context pointers!